File: Kalorie/Kalorie/Core/UseCases/MigrateAnonymousDataUseCase.swift
//
//  MigrateAnonymousDataUseCase.swift
//  Kalorie
//
//  Created by Josef Antoni on 07.08.2026.
//

import FirebaseAuth
import Foundation

protocol MigrateAnonymousDataUseCaseProtocol {
    func migrate(fromAnonymousUserId sourceUserId: String, credential: AuthCredential) async throws
    func resumeIfNeeded() async throws
}

struct MigrateAnonymousDataUseCase: MigrateAnonymousDataUseCaseProtocol {

    // MARK: - Properties

    private let dataProvider: any FirestoreDataProviderProtocol
    private let authProvider: any AuthProviderProtocol
    private let authCommandProvider: any AuthCommandProviderProtocol
    private let snapshotStore: any PendingMergeSnapshotStoreProtocol

    // MARK: - Init

    init(
        dataProvider: any FirestoreDataProviderProtocol,
        authProvider: any AuthProviderProtocol,
        authCommandProvider: any AuthCommandProviderProtocol,
        snapshotStore: any PendingMergeSnapshotStoreProtocol
    ) {
        self.dataProvider = dataProvider
        self.authProvider = authProvider
        self.authCommandProvider = authCommandProvider
        self.snapshotStore = snapshotStore
    }

    // MARK: - Functions

    func migrate(fromAnonymousUserId sourceUserId: String, credential: AuthCredential) async throws {
        let foodConsumed: [FoodConsumedDTO] = try await dataProvider.loadAsync(from: Constants.Firestore.foodConsumed(userId: sourceUserId))
        let favouriteFoods: [FavouriteFoodDTO] = try await dataProvider.loadAsync(from: Constants.Firestore.favouriteFoods(userId: sourceUserId))
        let snapshot = PendingMergeSnapshot(sourceAnonymousUserId: sourceUserId, foodConsumed: foodConsumed, favouriteFoods: favouriteFoods)
        try snapshotStore.save(snapshot)

        try await authCommandProvider.signIn(with: credential)

        try await writeAndCleanup(snapshot)
    }

    func resumeIfNeeded() async throws {
        guard let snapshot = try snapshotStore.load() else { return }
        guard let currentUserId = authProvider.userId else { return }

        if currentUserId == snapshot.sourceAnonymousUserId {
            try snapshotStore.delete()
        } else {
            try await writeAndCleanup(snapshot)
        }
    }

    // MARK: - Private

    private func writeAndCleanup(_ snapshot: PendingMergeSnapshot) async throws {
        guard let userId = authProvider.userId else { throw AuthError.notAuthenticated }
        let items = snapshot.foodConsumed.map { (item: $0, id: $0.id) }
        try await dataProvider.batchSetAsync(items, in: Constants.Firestore.foodConsumed(userId: userId))
        let favourites = snapshot.favouriteFoods.map { (item: $0, id: $0.id) }
        try await dataProvider.batchSetAsync(favourites, in: Constants.Firestore.favouriteFoods(userId: userId))
        try snapshotStore.delete()
    }
}

#if DEBUG
final class MigrateAnonymousDataUseCaseFake: MigrateAnonymousDataUseCaseProtocol {

    // MARK: - Properties

    var migrateError: Error?
    var resumeError: Error?
    private(set) var migrateCallCount = 0
    private(set) var lastSourceUserId: String?
    private(set) var resumeCallCount = 0

    // MARK: - Functions

    func migrate(fromAnonymousUserId sourceUserId: String, credential: AuthCredential) async throws {
        migrateCallCount += 1
        lastSourceUserId = sourceUserId
        if let migrateError { throw migrateError }
    }

    func resumeIfNeeded() async throws {
        resumeCallCount += 1
        if let resumeError { throw resumeError }
    }
}
#endif

File: Android/app/src/main/java/antoni/kalorie/core/usecases/MigrateAnonymousDataUseCase.kt
package antoni.kalorie.core.usecases

import antoni.kalorie.core.auth.AuthCommandProviderProtocol
import antoni.kalorie.core.auth.AuthError
import antoni.kalorie.core.auth.AuthProviderProtocol
import antoni.kalorie.core.auth.PendingMergeSnapshot
import antoni.kalorie.core.auth.PendingMergeSnapshotStoreProtocol
import antoni.kalorie.core.networking.FavouriteFoodDTO
import antoni.kalorie.core.networking.FirestoreDataProviderProtocol
import antoni.kalorie.core.networking.FoodConsumedDTO
import antoni.kalorie.core.networking.FoodItemPersonalPortionsDTO
import antoni.kalorie.core.networking.MyCreatedMealDTO
import antoni.kalorie.core.networking.batchSetAsync
import antoni.kalorie.core.networking.loadAsync
import antoni.kalorie.core.utils.Constants
import com.google.firebase.auth.AuthCredential
import kotlinx.coroutines.async
import kotlinx.coroutines.coroutineScope

interface MigrateAnonymousDataUseCaseProtocol {
    suspend fun migrate(sourceUserId: String, credential: AuthCredential)
    suspend fun resumeIfNeeded()
}

class MigrateAnonymousDataUseCase(
    private val dataProvider: FirestoreDataProviderProtocol,
    private val authProvider: AuthProviderProtocol,
    private val authCommandProvider: AuthCommandProviderProtocol,
    private val snapshotStore: PendingMergeSnapshotStoreProtocol,
) : MigrateAnonymousDataUseCaseProtocol {

    // MARK: - Functions

    override suspend fun migrate(sourceUserId: String, credential: AuthCredential) {
        val snapshot = coroutineScope {
            val foodConsumed = async { dataProvider.loadAsync<FoodConsumedDTO>(from = Constants.Firestore.foodConsumed(sourceUserId)) }
            val favouriteFoods = async { dataProvider.loadAsync<FavouriteFoodDTO>(from = Constants.Firestore.favouriteFoods(sourceUserId)) }
            val myCreatedMeals = async { dataProvider.loadAsync<MyCreatedMealDTO>(from = Constants.Firestore.myCreatedMeals(sourceUserId)) }
            val foodItemPortions = async {
                dataProvider.loadAsync<FoodItemPersonalPortionsDTO>(from = Constants.Firestore.foodItemPortions(sourceUserId))
            }
            PendingMergeSnapshot(
                sourceAnonymousUserId = sourceUserId,
                foodConsumed = foodConsumed.await(),
                favouriteFoods = favouriteFoods.await(),
                myCreatedMeals = myCreatedMeals.await(),
                foodItemPortions = foodItemPortions.await(),
            )
        }
        snapshotStore.save(snapshot)
        authCommandProvider.signIn(credential)
        writeAndCleanup(snapshot)
    }

    override suspend fun resumeIfNeeded() {
        val snapshot = snapshotStore.load() ?: return
        val currentUserId = authProvider.userId ?: return
        if (currentUserId == snapshot.sourceAnonymousUserId) {
            snapshotStore.delete()
        } else {
            writeAndCleanup(snapshot)
        }
    }

    // MARK: - Private

    private suspend fun writeAndCleanup(snapshot: PendingMergeSnapshot) {
        val userId = authProvider.userId ?: throw AuthError.NotAuthenticated
        dataProvider.batchSetAsync(snapshot.foodConsumed.map { it to it.id }, inCollection = Constants.Firestore.foodConsumed(userId))
        dataProvider.batchSetAsync(snapshot.favouriteFoods.map { it to it.id }, inCollection = Constants.Firestore.favouriteFoods(userId))
        dataProvider.batchSetAsync(snapshot.myCreatedMeals.map { it to it.id }, inCollection = Constants.Firestore.myCreatedMeals(userId))
        dataProvider.batchSetAsync(snapshot.foodItemPortions.map { it to it.id }, inCollection = Constants.Firestore.foodItemPortions(userId))
        snapshotStore.delete()
    }
}
